### PR TITLE
Environment Detection

### PR DESCRIPTION
**Pull request by jspectacular**
## Summary
firmware/utilities/build includes a shell script that detects the build environment. It works in both Windows and Linux.

## Test Plan
1. Build from firmware/ec. Verify that `firmware/utilities/build/build_ec_cfg.txt` is created and firmware builds with no new issues. The text file should be created even if the build fails.
2. Build from firmware/host. Verify that `firmware/utilities/build/build_host_cfg.txt` is created and host builds with no new issues. The text file should be created even if the build fails.
3. Build from firmware/ec/test.  Verify that `firmware/utilities/build/build_ut_cfg.txt` is created and firmware builds with no new issues. The text file should be created even if the build fails.

## Issues
Resolved issue #171

### CODE REVIEW

**Comment by mdlewisfb** [comment]
Can you add it to the unit tests? Have it grab the gcc version, and if possible the Unity version located in $(UNITY_ROOT)/release/version.info?
Can you also add a comment in the bug_repot.md template referring to this cfg output and asking for it to be attached?

**Comment by jspectacular** [comment]
[build_ec_cfg.txt](https://github.com/Telecominfraproject/OpenCellular/files/2560235/build_ec_cfg.txt)
[build_host_cfg.txt](https://github.com/Telecominfraproject/OpenCellular/files/2560236/build_host_cfg.txt)
[build_ut_cfg.txt](https://github.com/Telecominfraproject/OpenCellular/files/2560237/build_ut_cfg.txt)

Sample configuration files attached.

File: firmware/ec/test/Makefile [hunk 1]
```diff
@@ -87,9 +87,15 @@ endif
 
 # Make Targets
 # =============================================================================
-test: $(BUILD_PATHS) $(RESULTS)
+
+test: get_cfg ttest
+
+ttest: $(BUILD_PATHS) $(RESULTS)
```
Comment (mdlewisfb) [inline firmware/ec/test/Makefile]: I think you could just make test dependent on get_cfg finishing, if you want. i.e.
```
test: $(BUILD_PATHS) $(RESULT) get_cfg
```

**Comment by JoshuaJeyaraj** [comment]
I am not sure if this PR is the best place to suggest this comment .Currently the firmware directory is a bit hardcoded as below .

OCWARE_DIR ?= $(HOME)/OpenCellular/firmware/ec

There are times where we might be working on multiple issues at the same time . This restricts the build environment . Can we have 
OCWARE_DIR ?= $(shell pwd)

**Comment by jspectacular** [comment]
> If TIRTOS_DIR , XDCTOOLS_DIR and TOOLCHAIN is not present in the static path provided in the makefile can we add intelligence into this script to detect and update into this new path into the makefile.

Not sure what this suggestion means.  The script is already detecting if TIRTOS_DIR and TOOLCHAIN are valid paths (see lines 12 and 17 in envDetect.sh).  As for UNITY_ROOT, the build will just fail if you don't have Unity.

**Comment by jspectacular** [comment]
> I am not sure if this PR is the best place to suggest this comment .Currently the firmware directory is a bit hardcoded as below .
> 
> OCWARE_DIR ?= $(HOME)/OpenCellular/firmware/ec
> 
> There are times where we might be working on multiple issues at the same time . This restricts the build environment . Can we have
> OCWARE_DIR ?= $(shell pwd)

We can do it in this PR.

**Comment by jspectacular** [comment]
> > > If TIRTOS_DIR , XDCTOOLS_DIR and TOOLCHAIN is not present in the static path provided in the makefile can we add intelligence into this script to detect and update into this new path into the makefile.
> > 
> > 
> > Not sure what this suggestion means. The script is already detecting if TIRTOS_DIR and TOOLCHAIN are valid paths (see lines 12 and 17 in envDetect.sh). As for UNITY_ROOT, the build will just fail if you don't have Unity.
> 
> Will line if [ -d "$2" ]; search TIRTOS directory into HOME path or it can be from any path ?

If you had several toolchains and used find to search for them, how would you know which one you intend to build with?  The makefile always knows (or the build fails).  This is why the makefile passes the location of these directories to the script.

**Comment by jspectacular** [comment]
Are there outstanding issues with this PR?

**Pull request merged**